urld2fb1d2b05_thnaeff_Util_tgz-pJ8-ch_thn_util_tests_TreeTestJ8/src/ch/thn/util/thread/ControlledRunnable.java

```diff
@@ -0,0 +1,753 @@
+
+	/**
+	 * Returns <code>true</code> if the runnable is currently paused
+	 * 
+	 * @return
+	 */
+	public boolean isPaused() {
+		return paused;
+	}
+
+	/**
+	 * A flag which is set when the runnable has been requested to pause (with
+	 * {@link #pause(boolean) or {@link #pause(boolean, boolean)}) but it is still
+	 * running.
+	 * 
+	 * 
+	 * @return
+	 */
+	public boolean willPause() {
+		return pauseRequested && !paused;
+	}
+
+	// ========================================================================
+
+	/**
+	 * Blocks the execution for the given time or until the pause state of this
+	 * {@link ControlledRunnable} is changed (with reset/stop/pause). This is an
+	 * alternative for the {@link Runnable#wait(long)} method. It behaves exactly
+	 * like it, but this method here gets interrupted if reset or stop is called
+	 * and it puts the {@link ControlledRunnable} in the pause state. Using this
+	 * method is preferred because it does not lock the operation of the
+	 * {@link ControlledRunnable}.
+	 * 
+	 * @param timeout
+	 */
+	protected void controlledWait(long timeout) {
+		this.timeout = timeout;
+		runPause(true, true);
+	}
+
+	/**
+	 * Blocks the execution until the pause state of this {@link ControlledRunnable}
+	 * is changed (with reset/stop/pause). This is an alternative for the
+	 * {@link Runnable#wait()} method. It behaves exactly like it, but this
+	 * method here gets interrupted if reset or stop is called and it puts the
+	 * {@link ControlledRunnable} in the pause state.
+	 * 
+	 */
+	protected void controlledWait() {
+		controlledWait(0);
+	}
+
+	// ========================================================================
+
+
+	/**
+	 * Requests the runnable to stop.<br />
+	 * The result depends on the implementation, but usually this ends in a thread
+	 * to end.
+	 * 
+	 */
+	public void stop() {
+		stopRequested = true;
+
+		fireControlledRunnableListener(
+				new ControlledRunnableEvent(this, ControlledRunnableEvent.STATETYPE.RUNNING));
+
+		// Notify the pausing. runPause will exit if stop requested
+		synchronized (PAUSESYNC) {
+			PAUSESYNC.notify();
+		}
+
+	}
+
+	/**
+	 * Requests the runnable to stop. With flag to wait for the runnable to stop.<br />
+	 * The result depends on the implementation, but usually this ends in a thread
+	 * to end.
+	 * 
+	 * @param wait If set to <code>true</code>, this method blocks until the runnable has
+	 * reached its end point.
+	 */
+	public void stop(boolean wait) {
+		stopRequested = true;
+
+		fireControlledRunnableListener(
+				new ControlledRunnableEvent(this, ControlledRunnableEvent.STATETYPE.RUNNING));
+
+		// Notify the pausing. runPause will exit if stop requested
+		synchronized (PAUSESYNC) {
+			PAUSESYNC.notify();
+		}
+
+		if (!stopped) {
+			//Only synchronize if really necessary
+
+			synchronized (WAITFORSYNC) {
+				try {
+					//Wait for runnable to stop
+					while (!stopped) {
+						WAITFORSYNC.wait();
+					}
+				} catch (InterruptedException e) {}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)